Enable more ruff rules including `A`, `DTZ`, `PGH` and more

DanielYang59 opened the pull request:
## Summary

- Enable more `ruff` rules
- Revert an accidental change in tests for `io.lobster.inputs` in https://github.com/materialsproject/pymatgen/pull/3848#discussion_r1632144875
- Other format tweaks

### Enable more [`ruff` rules](https://docs.astral.sh/ruff/rules/)

The following requires code change:
- [rule family `PGH`](https://docs.astral.sh/ruff/rules/#pygrep-hooks-pgh) for stricter typing
- [rule family `A`](https://docs.astral.sh/ruff/rules/#flake8-builtins-a) to avoid shadowing of builtins

> [!CAUTION]
> b1aa5bbe8e and 50e9b7dde changed multiple `ValueError` to more semantic `TypeError` to fix [`TRY004`](https://docs.astral.sh/ruff/rules/type-check-without-type-error/). this is breaking. however, downstream repercussions are expected to be small.

pinging @mkhorton @shyuep for approval

#### Future PR
- 7fac0b466bbb18372334b87c990c46a1096b6415 need to fix #3791
- Enable rule family `S` and `NPY`

### Other format tweaks
- Replace `match.group(x)` with `match[x]`
- Replace `array[0:x]` slicing with `array[:x]`
- Replace already multi-lined `dict()` constructor with `{}` [for speed](https://switowski.com/blog/dict-function-vs-literal-syntax/) (almost 2x-4x speedup).

JaGeo commented:
@DanielYang59 could you trigger the atomate tests?

DanielYang59 commented:
@JaGeo Thanks for commenting. Can you let me how to achieve that? It should only be possible when pushing to `master` I guess?https://github.com/materialsproject/pymatgen/blob/624e6eb5cfefbf5736fe31b8910dd19c1042e94e/.github/workflows/test.yml#L94

JaGeo commented:
@DanielYang59 ah okay. Then, we will see after merging! Thanks!

DanielYang59 commented on pyproject.toml:
```diff
@@ -50,10 +59,13 @@ select = [
   "RET",     # flake8-return
   "RSE",     # flake8-raise
   "RUF",     # Ruff-specific rules
+  # "S",     # flake8-bandit
```
This [rule family](https://docs.astral.sh/ruff/rules/#flake8-bandit-s) provides some security related checks, it might be beneficial to enable it. But too many errors related to it, better to separate to another PR.

DanielYang59 commented on pyproject.toml:
```diff
@@ -25,20 +25,29 @@ line-length = 120
 [tool.ruff.lint]
 select = [
+  "A",       # flake8-builtins
+  # "ANN",   # flake8-annotations
```
This one is good, but it requires all arguments to be type annotated. We're not ready yet.

DanielYang59 commented on tests/analysis/chemenv/coordination_environments/test_read_write.py:
```diff
@@ -127,8 +127,6 @@ def test_structure_environments_neighbors_sets(self):
             str(nb_set) == "Neighbors Set for site #6 :\n - Coordination number : 4\n - Voronoi indices : 1, 4, 5, 6\n"
         )
 
-        assert nb_set == nb_set
```
It's so weird, there are multiple comparison with itself in 0ea568e48b9d0dbff7a8c39c43c048db41898500, does this serve any purpose that I'm not aware of?

janosh replied:
yes, these are there to check correctness of custom `__eq__` methods

janosh replied:
it might make sense to ignore `PLR0124` specifically in test files rather than ignore this error on a case by case basis

DanielYang59 replied:
Thanks, I was about to add it to `[tool.ruff.lint.per-file-ignores]` too :)

DanielYang59 commented:
@janosh Can you please advice on whether we want to fix [`A002`](https://docs.astral.sh/ruff/rules/builtin-argument-shadowing/)? It requires quite a lot of breaking argument renames.

Other than that, b1aa5bbe8e55461fa2e59672d5762272218f681b might have broken some unit test, and I would fix them :)

janosh commented:
> @janosh Can you please advice on whether we want to fix [A002](https://docs.astral.sh/ruff/rules/builtin-argument-shadowing/)? It requires quite a lot of breaking argument renames.

let's enable it to avoid builtin shadowing in the future but add `noqa: A002` for all existing cases to avoid breaking changes

DanielYang59 commented on pymatgen/vis/structure_vtk.py:
```diff
@@ -629,11 +629,11 @@ def add_faces(self, faces, color, opacity=0.35):
                 raise ValueError("Number of points for a face should be >= 3")
 
-    def add_edges(self, edges, type="line", linewidth=2, color=(0.0, 0.0, 0.0)):
+    def add_edges(self, edges, type="line", linewidth=2, color=(0.0, 0.0, 0.0)):  # noqa: A002
```
Does this unused `type` arg exist for a reason? If not, removing it would get rid of this `A002` error

janosh replied:
i'd guess it's there to conform with the signature of the parent class?

forgot to mention earlier, `ruff` [can auto-insert](https://docs.astral.sh/ruff/tutorial/#adding-rules) `noqa`s with the ``--add-noqa` flag

DanielYang59 replied:
> i'd guess it's there to conform with the signature of the parent class?

Not sure, this looks like the base class: https://github.com/materialsproject/pymatgen/blob/624e6eb5cfefbf5736fe31b8910dd19c1042e94e/pymatgen/vis/structure_vtk.py#L35

> forgot to mention earlier, `ruff` [can auto-insert](https://docs.astral.sh/ruff/tutorial/#adding-rules) `noqa`s with the `--add-noqa` flag

Thanks for the tip!

DanielYang59 commented:
@janosh Ready for review! Thanks for your time as always :)

janosh commented on tests/core/test_xcfunc.py:
```diff
@@ -14,7 +14,6 @@ def test_xcfunc_api(self):
         assert ixc_1.type == "LDA"
         assert ixc_1.name == "LDA_XC_TETER93"
-        assert ixc_1 == ixc_1
         assert ixc_1 == "LDA_XC_TETER93"
```
i think it makes sense to keep this line but move it into a separate method `def test_eq(self): ...`

in general, the name `test_xcfunc_api` is pretty meaningless and should be split into more descriptive method names, like `test_class_methods`, `test_serialization`, or something like that

DanielYang59 replied:
Done in 1fea2f43f02f189b6820ab604954d975b817dfae

janosh commented on tests/core/test_structure.py:
```diff
@@ -114,7 +114,6 @@ def test_as_dataframe(self):
     def test_equal(self):
         struct = self.struct
-        assert struct == struct
         assert struct == struct.copy()
```
these lines are important to keep in `__eq__` tests. we want to make sure the magic method returns `True` as expected

DanielYang59 replied:
Thanks, I didn't realize that!

janosh commented on tests/analysis/chemenv/coordination_environments/test_structure_environments.py:
```diff
@@ -151,8 +149,6 @@ def test_light_structure_environments(self):
 
-        assert nb_set == nb_set
-
         assert (
```
better to keep these as well though they should better be in separate test methods `def test_eq`, `test_equal` (the name of `__eq__` test methods is another thing to standardize across the code base)

DanielYang59 commented on tests/symmetry/test_analyzer.py:
```diff
@@ -263,76 +263,75 @@ def test_get_ir_reciprocal_mesh_map(self):
 
     def test_get_conventional_standard_structure(self):
         structure = Structure.from_file(f"{TEST_FILES_DIR}/cif/bcc_1927.cif")
-        assert structure == structure
```
Ping myself, revert this deletion

DanielYang59 commented:
@janosh I replaced some already multi-lined dictionary constructor `dict()` with `{}` for speed in 47d72f3c3c1738f7926ab30e30cceb6d4c8ffae4, I know you don't want to introduce extra line, so I just replaced those that wouldn't do so. 

From my side, the [speed up](https://switowski.com/blog/dict-function-vs-literal-syntax/) is something I really want to have (2x - 4x, though being in the nanosecond level) :)

DanielYang59 commented:
@shyuep Can I have your approval on the following potentially breaking change (change some `ValueError` to `TypeError` where the type doesn't match)? Thanks!

> [b1aa5bb](https://github.com/materialsproject/pymatgen/commit/b1aa5bbe8e55461fa2e59672d5762272218f681b) and [50e9b7d](https://github.com/materialsproject/pymatgen/commit/50e9b7ddee9764324795d84bff532ca6bc500ae1) changed multiple `ValueError` to more semantic `TypeError` to fix [`TRY004`](https://docs.astral.sh/ruff/rules/type-check-without-type-error/). this is breaking. however, downstream repercussions are expected to be small.

shyuep commented:
If the errors are not the type that are meant to be caught, it is fine to change the error type. What I mean is that if the error is something that happens because of a bad use of code. Rather than something like catching an expected exception case and then handling it in some way.

DanielYang59 commented:
Thanks for your reply!

Yes I could confirm those are exceptions we `raise` instead of catch (`try...except...`), see b1aa5bbe8e55461fa2e59672d5762272218f681b and 50e9b7ddee9764324795d84bff532ca6bc500ae1.

Mostly in the following pattern:
```diff
if not isinstance(variable, expected_type):
-    raise ValueError("variable should be in {expected_type}")
+    raise TypeError("variable should be in {expected_type}")
```

Still it could be breaking if any downstream operation reply on catching a `ValueError`, but I would consider this a fix as the wrong type of exception was raised in the first place.

shyuep commented:
Thanks I have merged them,.

DanielYang59 commented:
Thanks!
